extension/property.go

```diff
@@ -0,0 +1,84 @@
+package extension
+
+import (
+	"encoding/json"
+	"github.com/alibaba/sentinel-golang/logging"
+	"github.com/pkg/errors"
+	"hash/fnv"
+)
+
+var logger = logging.GetDefaultLogger()
+
+// PropertyConvert func is to convert source message bytes to the specific property.
+type PropertyConvert func(src []byte) interface{}
+
+// PropertyUpdate func is to update the specific properties to downstream.
+type PropertyUpdate func(data interface{}) error
+
+// abstract interface to
+type PropertyHandler interface {
+	// check whether the current src is consistent with last update property
+	isPropertyConsistent(src []byte) bool
+	// handle the current property
+	Handle(src []byte) error
+}
+
+// DefaultPropertyHandler encapsulate the Converter and updater of property.
+// One DefaultPropertyHandler instance is to handle one property type.
+// DefaultPropertyHandler should check whether current property is consistent with last update property
+// converter convert the message to the specific property
+// updater update the specific property to downstream.
+type DefaultPropertyHandler struct {
+	lastUpdatePropertyHash uint64
+
+	Convert PropertyConvert
+	Update  PropertyUpdate
+}
+
+func (h *DefaultPropertyHandler) isPropertyConsistent(src []byte) bool {
+	if len(src) == 0 {
+		return false
+	}
+	hash := fnv.New64()
+	_, err := hash.Write(src)
+	if err!=nil {
+		return false
+	}
+	currentHash := hash.Sum64()
+	// not thread safe, but it's ok.
+	if currentHash == h.lastUpdatePropertyHash {
+		return true
+	} else {
+		h.lastUpdatePropertyHash = currentHash
+		return false
+	}
+}
+
+func (h *DefaultPropertyHandler) Handle(src []byte) error {
+	if len(src) == 0 {
+		return h.Update(nil)
+	}
+	// convert to target property
+	realProperty := h.Convert(src)
+	if realProperty == nil {
+		return errors.New("The real property converted is nil.")
+	}
+
+	realSrc, err := json.Marshal(realProperty)
+	if err!=nil {
+		logger.Warnf("Fail to marshal property source, err:%+v.", err)
+	}
+	isConsistent := h.isPropertyConsistent(realSrc)
```

Comment:
Maybe we can compare the parsed entity directly, rather than convert it to JSON string then compare.